presto-main/src/main/java/io/prestosql/sql/ExpressionUtils.java

```diff
@@ -225,7 +232,33 @@ public static Expression combineDisjunctsWithDefault(Metadata metadata, Collecti
             return TRUE_LITERAL;
         }
 
-        return disjuncts.isEmpty() ? emptyDefault : or(disjuncts);
+        ImmutableMultimap.Builder<Expression, Expression> builder = ImmutableMultimap.builder();
+        ImmutableList.Builder<Expression> expressionBuilder = ImmutableList.builder();
+        for (Expression disjunct : disjuncts) {
+            if (disjunct instanceof ComparisonExpression && ((ComparisonExpression) disjunct).getOperator() == EQUAL) {
+                ComparisonExpression comparisonExpression = (ComparisonExpression) disjunct;
+                builder.put(comparisonExpression.getLeft(), comparisonExpression.getRight());
+            }
+            else if (disjunct instanceof InPredicate && ((InPredicate) disjunct).getValueList() instanceof InListExpression) {
+                InPredicate predicate = (InPredicate) disjunct;
+                ((InListExpression) predicate.getValueList()).getValues().forEach(value -> builder.put(predicate.getValue(), value));
+            }
+            else {
+                expressionBuilder.add(disjunct);
```

Comment:
This may end up reordering "cheap" predicates and put them behind more expensive ones.